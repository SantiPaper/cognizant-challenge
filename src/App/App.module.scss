.main{
  display: flex;
  justify-content: space-between;
  align-items: flex-start;
  padding: 1rem;
}

@media (max-width: 1000px) {
  .main{
    flex-direction: column;
    align-items: center;
  }
}
