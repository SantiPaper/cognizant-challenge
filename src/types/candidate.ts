export interface Candidate {
  id: string;
  name: string;
  comments: string;
  step: Step;
}

export type Step =
  | "Entrevista inicial"
  | "Entrevista técnica"
  | "Oferta"
  | "Asignación"
  | "Rechazo";
