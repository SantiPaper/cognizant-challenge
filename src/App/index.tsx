import React, {useEffect, useState} from "react";

import api from "../api/index";
import {Candidate, Step} from "../types/candidate";
import {Column} from "../components/Column";

import styles from "./App.module.scss";

const steps: Step[] = [
  "Entrevista inicial",
  "Entrevista técnica",
  "Oferta",
  "Asignación",
  "Rechazo",
];

const initialState = JSON.parse(localStorage.getItem("candidates") || "[]");

function App() {
  const [candidates, setCandidates] = useState<Candidate[]>(initialState);

  useEffect(() => {
    if (!candidates.length) {
      api.candidates.list().then((data) => {
        setCandidates(data);
      });
    }
  }, []);

  useEffect(() => {
    localStorage.setItem("candidates", JSON.stringify(candidates));
  }, [candidates]);

  const addCandidate = (candidato: Candidate) => {
    setCandidates([...candidates, candidato]);
  };

  const advance = (id: Candidate["id"], action: "avanzar" | "retroceder") => {
    const candidateIndex = candidates.findIndex((candidate) => candidate.id === id)!;
    const candidate = candidates[candidateIndex];
    const indexStep = steps.indexOf(candidate.step);

    const step = action === "avanzar" ? steps[indexStep + 1] : steps[indexStep - 1];

    if (step) {
      const newCandidates = [...candidates];

      candidate.step = step;
      newCandidates[candidateIndex] = candidate;
      setCandidates(newCandidates);
    }
  };

  return (
    <main className={styles.main}>
      {steps.map((step) => (
        <Column
          key={step}
          addCandidate={addCandidate}
          advance={advance}
          candidatos={candidates.filter((candidate) => candidate.step === step)}
          titulo={step}
        />
      ))}
    </main>
  );
}

export default App;
