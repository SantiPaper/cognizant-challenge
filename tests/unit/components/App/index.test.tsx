import React from "react";
import {getByRole, render, screen, waitFor} from "@testing-library/react";

import App from "../../../../src/App";
import {Candidate} from "../../../../src/types/candidate";
import api from "../../../../src/api";

const mockedCandidates: Candidate[] = [
  {
    id: "goncy",
    name: "Gonzalo Pozzo",
    step: "Entrevista tÃ©cnica",
    comments: "Medio pelo",
  },
  {
    id: "doe",
    name: "John Doe",
    step: "Entrevista inicial",
    comments: "",
  },
];

describe("Componente App", () => {
  const apiSpy = vi
    .spyOn(api.candidates, "list")
    .mockImplementation(() => Promise.resolve(mockedCandidates));
  const setItemSpy = vi.spyOn(localStorage, "setItem");

  test("Se renderiza correctamente", () => {
    const {container} = render(<App />);

    expect(container).toMatchSnapshot();
  });

  test("Setea localStorage", async () => {
    render(<App />);
    await waitFor(() => expect(screen.getByRole("heading", {name: "Gonzalo Pozzo"})));
    await waitFor(() =>
      expect(setItemSpy).toHaveBeenCalledWith("candidates", JSON.stringify(mockedCandidates)),
    );
  });
});
