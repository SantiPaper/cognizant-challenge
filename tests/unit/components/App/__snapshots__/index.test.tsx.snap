// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Componente App > Se renderiza correctamente 1`] = `
<div>
  <main
    class="_main_728e4a"
  >
    <div
      class="_column_545bf0"
    >
      <h2>
        Entrevista inicial
      </h2>
      <p>
        No hay candidatos
      </p>
      <button>
        Agregar candidato
      </button>
    </div>
    <div
      class="_column_545bf0"
    >
      <h2>
        Entrevista técnica
      </h2>
      <p>
        No hay candidatos
      </p>
    </div>
    <div
      class="_column_545bf0"
    >
      <h2>
        Oferta
      </h2>
      <p>
        No hay candidatos
      </p>
    </div>
    <div
      class="_column_545bf0"
    >
      <h2>
        Asignación
      </h2>
      <p>
        No hay candidatos
      </p>
    </div>
    <div
      class="_column_545bf0"
    >
      <h2>
        Rechazo
      </h2>
      <p>
        No hay candidatos
      </p>
    </div>
  </main>
</div>
`;
